### PR TITLE
made compile only dependencies provided as default gradle code

### PR DESCRIPTION
**Pull request by borissmidt**
Hey,

I really enjoy your library to do cross compilation on for scala projects and i think this should become the default scala plugin for gradle. Atleast how the feature is implemented.

At our company we where doing some work with cross compilation and graal and we found out that `compileOnly` dependencies are added to the pom as a `compile` dependency. This meant that our app included the full graal vm and failed durring build time.

The same thing applies to spark that the spark jar will be included in the fat jar which is not recommended.

So to resolve this i changed to code so compileOnly dependencies are added to the pom as `provided`.

I also added an entry in the README to show what types become what pom types. Because i was unable to override this behavior of your plugin.

I also did some boyscouting.

Kind regards, Boris Smidt

### CODE REVIEW

**Comment by borissmidt** [comment]
Would you be so kind to review this PR,
Because i will leave on holiday soon.

**Comment by prokod** [comment]
Hi Boris,

Thank you for the vote of confidence :)
I will review the PR after that but first I would like to comment in general.

The plugin already had this behavior where `compileOnly` was translated to `provided`
I have removed this functionality because of this [Gradle dependency management guideline](https://docs.gradle.org/current/userguide/building_java_projects.html?&_ga=2.209480082.102075081.1578311503-503725931.1575473283#sec:java_dependency_management_overview) which was updated when Gradle 5 got introduced back then.

An interesting discussion about that is [here](https://discuss.gradle.org/t/is-it-recommended-to-use-compileonly-over-implementation-if-another-module-use-implementation-already/26699)

You can see from that discussion this part is not fully baked from Gradle side and so this contributes mainly to why the plugin did not promote Gradle's `compileOnly` -> Maven's `provided` conversion.

**Comment by borissmidt** [comment]
I'm trying to put classes annotated with substratevm annotations, But this dependency should never be marked as compile in the maven pom. Since it will make any project including it fail.

It is the same thing with the spark library it should not be set as compile because it will then include it into the fat jar of the project.

I tried to override the pom manually, to mark this dependency as provided but there was no way to do this with your plugin.

So what do you propose i do? because it seems logical to me that the following conversions should be done:
compileOnly is provided
 implementation is runtime
compile/api is compile in the compile
annotationProcessor is not included in the pom.

**Comment by borissmidt** [comment]
If i run it in another project without the crossbuild plugin the compileOnly dependency is not added to the pom.xml of the library.

**Comment by prokod** [comment]
Hi Boris,

Couple of points (mixed reply/question to your latest comment and partly a review of your PR)

1. Regarding overriding pom this is a good feedback! and I will see how to tackle that so you can override this behavior.
2. Regarding `annotationProcessor` I did not cover that in tests - You are saying the plugin is not handling this configuration correctly right ?
3. I have seen you are approaching the Xml object in tests in a much nicer, more readable way 👍 
4. I looked at the build results here: [travis](https://travis-ci.org/prokod/gradle-crossbuild-scala/pull_requests) and I see this branch is failing. I have not added in the readme any section on contribution and how to go about that - my bad sorry, so let me just say the PR is failing on codenarc stage and both this and the full set of tests should pass. I will try to address the readme section on contribution soon enough.
5. Regarding the PR it is against master directly, I am using git flow process so this PR should be against `dev` branch

Regarding your last comment, compileOnly should not be added to pom.xml you are correct. This is a bug!

**Comment by borissmidt** [comment]
I tried `annotationProcessor`, but this didn't allow me to use the annotations of the `annotationProcessor` dependency so i couldn't use it. I think is only for enabling an annotation processor like lombok.

> Regarding your last comment, compileOnly should not be added to pom.xml you are correct. This is a bug!

Then i can easily update this PR so it removes the `compileOnly` dependency.

**Comment by borissmidt** [comment]
I've changed the code so it just ignores the compileOnly dependencies.

**Comment by borissmidt** [comment]
I changed the pr to merge with dev and i will resolve the merge conflicts later today.

**Comment by prokod** [comment]
Hi Boris,
You should rebase your branch on top of dev. Right now it is very hard to review this as the merge strategy brings in too much noise.

**Comment by borissmidt** [comment]
I took some time to rebase it.

**Comment by borissmidt** [comment]
accidentaly closed it.

**Comment by prokod** [comment]
Hi Boris I have not updated dev to include latest documentation from hotfix branch so I have fixed that.
Please try rebasing I did that for your branch locally and it went fine (not straight away) though

File: src/main/groovy/com/github/prokod/gradle/crossbuild/tasks/AbstractCrossBuildPomTask.groovy [hunk 1]
```diff
@@ -30,6 +30,7 @@ abstract class AbstractCrossBuildPomTask extends DefaultTask {
     static enum ScopeType {
         COMPILE,
         RUNTIME,
-        PROVIDED
+        PROVIDED,
+        COMPILE_ONLY
```
Comment (prokod) [inline src/main/groovy/com/github/prokod/gradle/crossbuild/tasks/AbstractCrossBuildPomTask.groovy:34]: Why did you add COMPILE_ONLY scope ? those enumerations are strictly for maven side. So now it doesn't make sense

**Comment by borissmidt** [comment]
So it was consistent with the rest of the code. You could rename it to
NOT_IN_POM.

File: src/main/groovy/com/github/prokod/gradle/crossbuild/tasks/CrossBuildPomTask.groovy [hunk 2]
```diff
@@ -39,52 +42,82 @@ class CrossBuildPomTask extends AbstractCrossBuildPomTask {
...
-                case ScopeType.COMPILE:
-                    return project.configurations[crossBuildSourceSet.compileClasspathConfigurationName]
-                case ScopeType.RUNTIME:
-                    return project.configurations[crossBuildSourceSet.runtimeClasspathConfigurationName]
+        def resolveSourceSet = { ScopeType scope ->
+            def sourceSetName = {
+                switch (scope) {
+                    case ScopeType.COMPILE:
+                        return crossBuildSourceSet.compileClasspathConfigurationName
+                    case ScopeType.COMPILE_ONLY:
```
Comment (borissmidt) [inline src/main/groovy/com/github/prokod/gradle/crossbuild/tasks/CrossBuildPomTask.groovy:50]: then it was still consistent with this code.
I think there should be a scope type which tells do not include it.

**Comment by prokod** [comment]
I had some more time (Sorry for the inconvenience and delays) and I found out there is an issue with the original code as the compileOnly dependencies should not have appeared in the first place when assembling dependencies for the pom (`compileClasspath`) I am investigating further and I will keep you updated.
Thx for the time you took to compile this PR, it is important as it surfaced some inherent issues with the plugin I would like to tackle.
I am not sure if in the end of the day this PR will get merged because this is not the way I would like the code for this plugin to be evolving, but it is certainly a milestone in the project, a catalyst which raised my awareness for contribution aspects and more.

**Comment by borissmidt** [comment]
Yes, as long as the plugin works i would be happy, you can always take the
pieces you like. For example the unit test and maybe the deduplication of
some code.
